### PR TITLE
Swift: Wire initializer refactor

### PR DESCRIPTION
**Pull request by lickel**
This removes the member-wise initializer by default.

There's two reasons for this:
* This is believed to cause code bloat
* This will have unexpected behaviors as we better handle proto extensions

Instead of a member wise initializer, this exposes a configuration closure for any optional parameters.
There's still guaranteed type safety as all `required` parameters _must_ have initialized values.

### Note

Memberwise initializers should now be considered deprecated and will be removed in the near future (for example, November 2023)

### TODO:

- [x] Add deprecation annotations
- [x] Remove `configure` when it's irrelevant

### CODE REVIEW

**Review by dnkoutso — APPROVED** [review]

**Comment by oldergod** [comment]
`./gradlew spotlessApply`

File: wire-runtime-swift/src/main/swift/wellknowntypes/Duration.swift [hunk 1]
```diff
@@ -81,9 +81,14 @@ public struct Duration {
     public var nanos: Int32
     public var unknownFields: Foundation.Data = .init()
 
-    public init(seconds: Int64, nanos: Int32) {
+    public init(
+        seconds: Int64,
+        nanos: Int32,
+        configure: (inout Self) -> Void = { _ in }
```
Comment (efirestone) [inline wire-runtime-swift/src/main/swift/wellknowntypes/Duration.swift:87]: Is there value in having the `configure` block if all of the fields are required? (Or if there are no fields, like the `OneOfOptions` below?)

File: wire-runtime-swift/src/test/swift/sample/Dinosaur.swift [hunk 2]
```diff
@@ -12,17 +12,27 @@ public struct Dinosaur {
...
+    public init(configure: (inout Self) -> Void = { _ in }) {
+        configure(&self)
+    }
+
+}
+
+#if WIRE_INCLUDE_MEMBERWISE_INITIALIZER
+extension Dinosaur {
+
+    @_disfavoredOverload
```
Comment (efirestone) [inline wire-runtime-swift/src/test/swift/sample/Dinosaur.swift:30]: Fancy!

File: wire-swift-generator/src/main/java/com/squareup/wire/swift/SwiftGenerator.kt [hunk 3]
```diff
@@ -255,7 +259,7 @@ class SwiftGenerator private constructor(
     fileMembers: MutableList<FileMemberSpec>,
   ): List<TypeSpec> {
     val structType = type.typeName
-    val oneOfEnumNames = type.oneOfs.associateWith { structType.nestedType(it.name.capitalize(US)) }
```
Comment (efirestone) [inline wire-swift-generator/src/main/java/com/squareup/wire/swift/SwiftGenerator.kt:258]: Why this change?

File: wire-swift-generator/src/main/java/com/squareup/wire/swift/SwiftGenerator.kt [hunk 4]
```diff
@@ -965,38 +950,165 @@ class SwiftGenerator private constructor(
...
+          )
+        }
+        .build(),
+    )
+
+    if (type.fields.all { it.isRequiredParameter } && type.oneOfs.isEmpty()) {
+      return
+    }
+
+    val memberwiseExtension = ExtensionSpec.builder(structType)
```
Comment (efirestone) [inline wire-swift-generator/src/main/java/com/squareup/wire/swift/SwiftGenerator.kt:1014]: Maybe add a comment noting we want to remove this, but it's left in temporarily for compatibility.

File: wire-tests-swift/src/main/swift/Form.swift [hunk 5]
```diff
@@ -180,13 +189,26 @@ public struct Form {
 
         public var unknownFields: Foundation.Data = .init()
 
-        public init() {
+        public init(configure: (inout Self) -> Void = { _ in }) {
```
Comment (efirestone) [inline wire-tests-swift/src/main/swift/Form.swift:192]: Yeah, we should remove the `configure` if there are no params.

**Review by efirestone — APPROVED** [review]

File: wire-tests-swift/src/main/swift/Form.swift (hunk 5 above)
Comment (lickel) [reply wire-tests-swift/src/main/swift/Form.swift:192]: - [ ] Remove Configure when appropriate

File: wire-swift-generator/src/main/java/com/squareup/wire/swift/SwiftGenerator.kt (hunk 4 above)
Comment (lickel) [reply wire-swift-generator/src/main/java/com/squareup/wire/swift/SwiftGenerator.kt:1014]: - [ ] Add deprecation annotations

File: wire-runtime-swift/src/test/swift/sample/Dinosaur.swift (hunk 2 above)
Comment (lickel) [reply wire-runtime-swift/src/test/swift/sample/Dinosaur.swift:30]: hehe I realized I would need it immediately for a bunch of types 🙂
